File: backend/src/routes/color_card.rs
//! 色卡仓储管理路由
//!
//! V15 P0-F03~F05 重构：删除借出/归还(borrow)路由，新增发放(issue)路由
//! 16 端点：色卡 CRUD + 色号 CRUD + 发放/归还/遗失/损坏/取消/列表/详情 + 扫码/导入/导出

use axum::{
    routing::{get, post, put},
    Router,
};

use crate::container::AppState;
use crate::handlers::color_card;
use crate::handlers::print_handler;

/// 色卡仓储管理路由（nest 到 /api/v1/erp/color-cards）
pub fn routes() -> Router<AppState> {
    Router::new()
        // 色卡 CRUD
        .route(
            "/",
            get(color_card::list_color_cards)
                .post(color_card::create_color_card),
        )
        .route(
            "/:id",
            get(color_card::get_color_card)
                .put(color_card::update_color_card)
                .delete(color_card::archive_color_card),
        )
        // 直接标记色卡为遗失（不同于发放记录遗失 /issues/:record_id/lost）
        .route(
            "/:id/mark-lost",
            post(color_card::mark_card_lost),
        )
        // 色号 CRUD
        .route(
            "/:id/items",
            get(color_card::list_color_items)
                .post(color_card::create_color_item),
        )
        .route(
            "/:id/items/batch",
            post(color_card::batch_import_items),
        )
        .route(
            "/:id/items/:item_id",
            put(color_card::update_color_item)
                .delete(color_card::delete_color_item),
        )
        // V15 P0-F05：发放 / 归还 / 遗失 / 损坏 / 取消（替代旧 borrow 路由）
        .route("/issues", post(color_card::issue_color_card))
        .route(
            "/issues",
            get(color_card::list_issues),
        )
        .route(
            "/issues/:record_id",
            get(color_card::get_issue),
        )
        .route(
            "/issues/:record_id/return",
            post(color_card::return_issue),
        )
        .route(
            "/issues/:record_id/lost",
            post(color_card::mark_issue_lost),
        )
        .route(
            "/issues/:record_id/damaged",
            post(color_card::mark_issue_damaged),
        )
        .route(
            "/issues/:record_id/cancel",
            post(color_card::cancel_issue),
        )
        // V15 P1-08-7：色卡发放记录导出 xlsx
        .route("/issues/export", get(color_card::export_issue_records))
        // V15 P2 类九 10.3-3：色卡发放报表（发放明细/汇总/客户台账/过期未使用/订单关联）
        .route(
            "/reports/issue-detail",
            get(color_card::issue_detail_report),
        )
        .route(
            "/reports/issue-detail/export",
            get(color_card::export_issue_detail_report),
        )
        .route(
            "/reports/issue-summary",
            get(color_card::issue_summary_report),
        )
        .route(
            "/reports/customer-ledger/:customer_id",
            get(color_card::customer_color_card_ledger),
        )
        .route(
            "/reports/expired-unused",
            get(color_card::expired_unused_report),
        )
        .route(
            "/reports/order-related/:sales_order_id",
            get(color_card::order_related_report),
        )
        // V15 P2 类九 10.5-2：库存预警
        .route("/warnings", get(color_card::check_all_warnings))
        .route(
            "/warnings/:color_card_id",
            get(color_card::check_single_warning),
        )
        // V15 P2 类九 10.3-4：成本核算（制作成本归集/发放成本结转/取消恢复/过期损失）
        .route(
            "/cost/production/:color_card_id",
            get(color_card::collect_production_cost),
        )
        .route(
            "/cost/issue/:record_id/transfer",
            get(color_card::transfer_issue_cost),
        )
        .route(
            "/cost/issue/:record_id/restore",
            post(color_card::restore_cost_on_cancel),
        )
        .route(
            "/cost/issue/:record_id/expiry-loss",
            get(color_card::calculate_expiry_loss),
        )
        // V15 P2 类九 10.5-3：发放统计（每日日报）
        .route(
            "/statistics/daily",
            get(color_card::generate_daily_stats),
        )
        // 扫码查询
        .route("/scan/:code", get(color_card::scan_color_code))
        // 按 ID 扫码查询
        .route("/scan-by-id/:id", get(color_card::scan_color_by_id))
        // 导出 CSV
        .route(
            "/export/:id",
            get(color_card::export_color_card),
        )
        // 打印路由
        .route(
            "/:id/bulk-approval/print",
            get(print_handler::bulk_color_approval_print_docx),
        )
        .route(
            "/:id/issue/print",
            get(print_handler::color_card_issue_print_docx),
        )
        .route(
            "/:id/lab-dip/print",
            get(print_handler::lab_dip_request_print_docx),
        )
}

File: frontend/src/api/color-card.ts
// 色卡仓储管理 API 客户端
// 16 端点封装
// 创建时间: 2026-06-17
// V15 P0-F07 重构（2026-07-17）：删除 borrow 模式，替换为 issue 发放模式
// 端点路径相对于 baseURL（/api/v1/erp），不要重复添加前缀，否则会产生双重前缀

import { request } from './request'

// ============== 枚举与常量 ==============

// 色卡类型
export const COLOR_CARD_TYPE = {
  PANTONE: 'PANTONE',
  CNCS: 'CNCS',
  CUSTOM: 'CUSTOM',
} as const

export const COLOR_CARD_TYPE_LABELS: Record<string, string> = {
  PANTONE: 'PANTONE',
  CNCS: 'CNCS',
  CUSTOM: '自定义',
}

// 色卡状态
export const COLOR_CARD_STATUS = {
  active: '在用',
  archived: '已归档',
  lost: '遗失',
} as const

export const COLOR_CARD_STATUS_COLORS: Record<string, string> = {
  active: 'success',
  archived: 'info',
  lost: 'danger',
}

// 发放状态（V15 P0-F07：替代原 BORROW_STATUS，borrow 模式已废弃）
export const ISSUE_STATUS = {
  issued: '发放中',
  returned: '已归还',
  lost: '遗失',
  damaged: '损坏',
  cancelled: '已取消',
} as const

export const ISSUE_STATUS_COLORS: Record<string, string> = {
  issued: 'warning',
  returned: 'success',
  lost: 'danger',
  damaged: 'danger',
  cancelled: 'info',
}

// 季节标签
export const SEASON_LABELS: Record<string, string> = {
  '2024SS': '2024 春夏',
  '2024AW': '2024 秋冬',
  '2025SS': '2025 春夏',
  '2025AW': '2025 秋冬',
  '经典': '经典款',
}

// ============== 类型定义 ==============

export interface ColorCardListItem {
  id: number
  card_no: string
  card_name: string
  card_type: string
  season?: string
  brand?: string
  total_colors: number
  status: string
  cover_image_url?: string
  // V15 P0-F10：色卡总库存数量
  stock_quantity: number
  // V15 P0-F10：已发放数量（可用 = stock_quantity - issued_quantity）
  issued_quantity: number
  created_at: string
}

export interface ColorItemInfo {
  id: number
  color_code: string
  color_name: string
  rgb_r: number
  rgb_g: number
  rgb_b: number
  cmyk_c?: number
  cmyk_m?: number
  cmyk_y?: number
  cmyk_k?: number
  lab_l?: number
  lab_a?: number
  lab_b?: number
  pantone_code?: string
  cncs_code?: string
  custom_code?: string
  hex_value: string
  dye_recipe_id?: number
  product_color_price_id?: number
  swatch_image_url?: string
  sequence: number
}

export interface ColorCardDetail extends ColorCardListItem {
  description?: string
  items: ColorItemInfo[]
  updated_at: string
}
// 发放记录信息（V15 P0-F07：替代原 BorrowRecordInfo，borrow 模式已废弃）
// 与后端 handlers/color_card/issue.rs::IssueRecordInfo 字段对齐
export interface IssueRecordInfo {
  id: number
  color_card_id: number
  customer_id: number
  issue_qty: number
  issued_by: number
  issued_at: string
  expected_return_date?: string
  actual_return_date?: string
  status: string
  purpose?: string
  remark?: string
  compensation_amount?: number
  returned_by?: number
  dye_lot_no?: string
  created_at: string
  updated_at: string
}

export interface PagedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

// ============== 色卡 CRUD ==============

export function listColorCards(params: {
  page?: number
  page_size?: number
  card_type?: string
  season?: string
  status?: string
  keyword?: string
}) {
  return request.get<{ data: PagedResponse<ColorCardListItem> }>('/color-cards', { params })
}

export function getColorCard(id: number) {
  return request.get<{ data: ColorCardDetail }>(`/color-cards/${id}`)
}

export function createColorCard(dto: {
  card_no: string
  card_name: string
  card_type: string
  season?: string
  brand?: string
  description?: string
  cover_image_url?: string
}) {
  return request.post<{ data: ColorCardListItem }>('/color-cards', dto)
}

export function updateColorCard(id: number, dto: Partial<{
  card_name: string
  card_type: string
  season: string
  brand: string
  description: string
  cover_image_url: string
}>) {
  return request.put<{ data: ColorCardListItem }>(`/color-cards/${id}`, dto)
}

export function archiveColorCard(id: number, reason?: string) {
  return request.delete<{ data: ColorCardListItem }>(`/color-cards/${id}`, {
    data: { reason },
  })
}

// ============== 色号 CRUD ==============

export function listColorItems(cardId: number, params?: { page?: number; page_size?: number }) {
  return request.get<{ data: PagedResponse<ColorItemInfo> }>(
    `/color-cards/${cardId}/items`,
    { params },
  )
}

export function createColorItem(cardId: number, dto: Partial<ColorItemInfo>) {
  return request.post<{ data: ColorItemInfo }>(
    `/color-cards/${cardId}/items`,
    dto,
  )
}

export function updateColorItem(cardId: number, itemId: number, dto: Partial<ColorItemInfo>) {
  return request.put<{ data: ColorItemInfo }>(
    `/color-cards/${cardId}/items/${itemId}`,
    dto,
  )
}

export function deleteColorItem(cardId: number, itemId: number) {
  return request.delete<{ data: null }>(
    `/color-cards/${cardId}/items/${itemId}`,
  )
}

// 批次 98 P2-D 修复（v5 复审）：原 errors: any[] 改为显式错误项接口
export interface BatchImportError {
  row: number
  message: string
  field?: string
}

export function batchImportItems(cardId: number, items: Partial<ColorItemInfo>[]) {
  return request.post<
    { data: { success_count: number; failed_count: number; errors: BatchImportError[]; total_colors: number } }
  >(
    `/color-cards/${cardId}/items/batch`,
    { items },
  )
}

// ============== 发放管理（V15 P0-F07：替代原借出管理，borrow 模式已废弃）==============

// 创建发放记录 DTO（与后端 handlers/color_card/issue.rs::IssueColorCardDto 对齐）
export function issueColorCard(dto: {
  color_card_id: number
  customer_id: number
  issue_qty?: number
  expected_return_date?: string
  purpose?: string
  remark?: string
  dye_lot_no?: string
}) {
  return request.post<{ data: IssueRecordInfo }>('/color-cards/issues', dto)
}

// 归还色卡（与后端 ReturnColorCardDto 对齐）
export function returnIssue(recordId: number, dto: { actual_return_date?: string; remark?: string }) {
  return request.post<{ data: IssueRecordInfo }>(
    `/color-cards/issues/${recordId}/return`,
    dto,
  )
}

// 登记遗失（与后端 MarkLostDto 对齐）
export function markIssueLost(recordId: number, dto: { compensation_amount: number; remark?: string }) {
  return request.post<{ data: IssueRecordInfo }>(
    `/color-cards/issues/${recordId}/lost`,
    dto,
  )
}

// 标记损坏（与后端 MarkDamagedDto 对齐）
export function markIssueDamaged(recordId: number, dto: { compensation_amount?: number; remark?: string }) {
  return request.post<{ data: IssueRecordInfo }>(
    `/color-cards/issues/${recordId}/damaged`,
    dto,
  )
}

// 取消发放（与后端 CancelIssueDto 对齐）
export function cancelIssue(recordId: number, dto: { remark?: string }) {
  return request.post<{ data: IssueRecordInfo }>(
    `/color-cards/issues/${recordId}/cancel`,
    dto,
  )
}

// 发放记录详情
export function getIssue(recordId: number) {
  return request.get<{ data: IssueRecordInfo }>(`/color-cards/issues/${recordId}`)
}

// 发放记录列表（与后端 ListIssuesQuery 对齐）
export function listIssues(params: {
  color_card_id?: number
  customer_id?: number
  status?: string
  page?: number
  page_size?: number
  from_date?: string
  to_date?: string
}) {
  return request.get<{ data: PagedResponse<IssueRecordInfo> }>(
    '/color-cards/issues',
    { params },
  )
}

// ============== 扫码查询 ==============

// P2-9c 修复（批次 82 v1 复审）：扫码查询返回类型 any → unknown
export function scanColorCode(code: string) {
  return request.get<{ data: unknown }>(`/color-cards/scan/${encodeURIComponent(code)}`)
}

export function exportColorCardUrl(cardId: number) {
  return `/color-cards/export/${cardId}`
}
